molecule/provisioners/libvirtprovisioner.py

```diff
@@ -0,0 +1,923 @@
+        devices = ET.SubElement(dom, 'devices')
+
+        # Disk elements
+        disk = ET.SubElement(devices, 'disk', type='file', device='disk')
+        ET.SubElement(
+            disk, 'driver', name='qemu', type='qcow2', cache='default')
+        ET.SubElement(
+            disk,
+            'source',
+            file=(os.path.join(self._pool_path, instance['name'] + '.img')))
+        ET.SubElement(disk, 'target', dev='vda', bus='virtio')
+
+        # Serial/console/usb elements
+        serial = ET.SubElement(devices, 'serial', type='pty')
+        target = ET.SubElement(serial, 'target', port='0')
+        alias = ET.SubElement(serial, 'alias', name='serial0')
+
+        console = ET.SubElement(devices, 'console', type='pty')
+        target = ET.SubElement(console, 'target', port='0', type='serial')
+        alias = ET.SubElement(console, 'alias', name='serial0')
+
+        # Network interface elements
+        #iface = ET.SubElement(devices, 'interface', type='network')
+        #ET.SubElement(iface, 'source', network='default')
+        for nic in instance.get('interfaces', [{'network_name': 'default'}]):
+            iface = ET.SubElement(devices, 'interface', type='network')
+            if 'address' in nic:
+                ip = netaddr.IPAddress(nic['address'])
+                # Find the libvirt network for this interface is connected to
+                cidr = netaddr.IPNetwork(
+                    filter(lambda net: net['name'] == nic['network_name'],
+                           self.molecule.config.config['libvirt']['networks'])[
+                               0]['cidr'])
+                ET.SubElement(
+                    iface,
+                    'ip',
+                    address=nic['address'],
+                    prefix=str(cidr.prefixlen))
+                ET.SubElement(
+                    iface,
+                    'route',
+                    family='ipv4',
+                    address=str(cidr.network),
+                    prefix=str(cidr.prefixlen),
+                    gateway=str(cidr.ip))
+            ET.SubElement(iface, 'source', network=nic['network_name'])
+            ET.SubElement(iface, 'model', type='virtio')
+        # Finally
+        domxml = ET.tostring(dom)
+        LOG.debug("\tXMLDesc: {}".format(domxml))
+        return domxml
+
+    def _fetch(self, url, filename):
+        """
+        Fetches a URL, saves response to <filename>.
+
+        :param: URL to fetch
+        :param: destination filename, will be in either self._sources_path or self._pool_path
+        """
+        if filename.endswith('.box'):
+            path = self._sources_path
+        else:
+            path = self._pool_path
+        utilities.print_info("Fetching image {} \n\t to {}...".format(
+            url, os.path.join(path, filename)))
+        r = requests.get(url, stream=True)
+        if r.status_code != 200:
+            try:
+                os.unlink(os.path.join(path, filename))
```

Comment:
Why are you removing it if the download fails?  Shouldn't we just raise and be done with it?
`s/unlink/remove` (this is Python, not bash :))